src/main/java/me/confor/velocity/chat/modules/GlobalChat.java

```diff
@@ -66,57 +64,61 @@ public void onPlayerChat(PlayerChatEvent event) {
             msg = msg.replaceText(config.urlReplacement);
 
         if (config.GLOBAL_CHAT_PASSTHROUGH)
-            sendMessage(msg,currentServer.get().getServer());
-        else
             sendMessage(msg);
+        else
+            sendMessage(msg, currentServer.get().getServer());
 
         if (config.GLOBAL_CHAT_TO_CONSOLE)
             this.logger.info("GLOBAL: <{}> {}", player, message);
 
-        if (!config.GLOBAL_CHAT_PASSTHROUGH)
+        /*
+        broken >=1.19.1
+
+        if (config.GLOBAL_CHAT_PASSTHROUGH)
             event.setResult(ChatResult.denied());
+         */
     }
 
     @Subscribe
-    public void onConnect(LoginEvent event) {
-        if (!config.JOIN_ENABLE)
-            return;
+    public void onConnect(ServerConnectedEvent event) {
+        RegisteredServer currentServer = event.getServer();
+        Optional<RegisteredServer> previousServer = event.getPreviousServer();
 
         String player = event.getPlayer().getUsername();
+        String server = currentServer.getServerInfo().getName();
 
-        Component msg = parseMessage(config.JOIN_FORMAT, List.of(
-                new ChatTemplate("player", player, false)
-        ));
-
-        sendMessage(msg);
-    }
-
-    @Subscribe
-    public void onDisconnect(DisconnectEvent event) {
-        if (!config.QUIT_ENABLE)
-            return;
+        Component msg;
 
-        Optional<ServerConnection> currentServer = event.getPlayer().getCurrentServer();
-        if (currentServer.isEmpty())
-            return;
+        if (previousServer.isPresent()) {
+            if (!config.SWITCH_ENABLE)
+                return;
 
-        String player = event.getPlayer().getUsername();
-        String server = currentServer.get().getServerInfo().getName();
+            msg = parseMessage(config.SWITCH_FORMAT, List.of(
+                new ChatTemplate("player", player, false),
+                new ChatTemplate("server", server, false),
+                new ChatTemplate("previous_server", previousServer.get().getServerInfo().getName(), false)
+            ));
+        } else {
+            if (!config.JOIN_ENABLE)
+                return;
 
-        Component msg = parseMessage(config.QUIT_FORMAT, List.of(
+            msg = parseMessage(config.JOIN_FORMAT, List.of(
                 new ChatTemplate("player", player, false),
                 new ChatTemplate("server", server, false)
-        ));
+            ));
 
-        if (config.GLOBAL_CHAT_PASSTHROUGH)
-            sendMessage(msg,currentServer.get().getServer());
-        else
-            sendMessage(msg);
+            if (!config.GLOBAL_CHAT_PASSTHROUGH) {
+                sendMessage(msg, currentServer);
+                return;
+            }
+        }
+
+        sendMessage(msg);
```

Comment:
You can't see your own join messages in vanilla Minecraft, but switch messages make sense.

I guess it's better to just show both to any joining player, rather than adhering to vanilla behavior and implementing separate handling just for switch messages.

Reply:
> You can't see your own join messages in vanilla Minecraft

I could when I was testing, at least on 1.19.4

Reply:
From my testing, no chat messages are sent to my vanilla 1.19.4 client when first joining a vanilla 1.19.4 server.

<img width="526" alt="Screenshot 2023-05-30 at 8 10 19 AM" src="https://github.com/confor/velocity-chat/assets/11802133/cd1d3b48-8289-4a7f-823d-d9ed21d12eeb">